core/src/main/java/com/linecorp/armeria/server/logging/structured/StructuredLoggingService.java

```diff
@@ -0,0 +1,112 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.logging.structured;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.logging.MessageLogConsumer;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.ResponseLog;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerListener;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceConfig;
+import com.linecorp.armeria.server.logging.LogCollectingService;
+
+/**
+ * A decorating service which provides support of structured and optionally externalized request/response
+ * content logging.
+ *
+ * @param <I> the {@link Request} type
+ * @param <O> the {@link Response} type
+ * @param <L> the type of the structured log representation
+ */
+public class StructuredLoggingService<I extends Request, O extends Response, L>
+        extends LogCollectingService<I, O> {
+
+    public static <I extends Request, O extends Response, L>
+    Function<Service<? super I, ? extends O>,
+            StructuredLoggingService<I, O, L>> newDecorator(StructuredLogWriter<L> logWriter,
+                                                            StructuredLogBuilder<L> logBuilder) {
+        return service -> new StructuredLoggingService<>(service, logWriter, logBuilder);
+    }
+
+    private final StructuredLogWriter<L> logWriter;
+
+    /**
+     * Creates a new {@link StructuredLoggingService}
+     *
+     * @param delegate the {@link Service} being decorated
+     * @param logWriter an instance of {@link StructuredLogWriter} which is used to write a structured logs
+     * @param logBuilder an instance of {@link StructuredLogBuilder} which is used to construct an entry of
+     * structured log
+     */
+    private StructuredLoggingService(Service<? super I, ? extends O> delegate,
+                                     StructuredLogWriter<L> logWriter,
+                                     StructuredLogBuilder<L> logBuilder) {
+        super(delegate, new StructuredLoggingConsumer<>(logWriter, logBuilder));
+        this.logWriter = logWriter;
+    }
+
+    @Override
+    public void serviceAdded(ServiceConfig cfg) throws Exception {
+        super.serviceAdded(cfg);
+
```

Comment:
hmm, maybe it's already not likely to happen that the `ServiceLoggingService` instance added to more than once to different servers since we made a constructor of it private and the `newDecorator` is the only way(which returns a function that instantiates `ServiceLoggingService` each time it called) to instantiate a service?

and the essential problem here is to avoid the same `logWriter` instances being used by two different servers and being closed more than once? then this definition still doesn't prevent that happen? no?